core/src/main/java/hudson/model/Queue.java

```diff
@@ -1691,6 +1691,17 @@ private Runnable makeFlyWeightTaskBuildable(final BuildableItem p){
         //we double check if this is a flyweight task
         if (p.task instanceof FlyweightTask) {
             Jenkins h = Jenkins.getInstance();
+
+            Label lbl = p.getAssignedLabel();
+
+            if (lbl != null && lbl.equals(h.getSelfLabel())) {
+                if (h.canTake(p) == null) {
+                    return createFlyWeightTaskRunnable(p, h.toComputer());
+                } else {
+                    return null;
```

Comment:
This seems to result in a regression when another node has the `master` label assigned manually, and master itself is offline. Previously, the task would be scheduled on the other node, now it won't be scheduled. WDYT?

Reply:
> when another node has the `master` label assigned manually

Is that even legal? Should probably be blocked.

Reply:
Most `selfLabel` can be used as a regular label (and for agent self-labels, there's no way to really prevent this, unless you ignore all manually assigned labels that are an agent's self-label, which then makes any node's labels dependent on all other nodes in the system, not great either). Namespacing these as `a:name` and `l:whatever` would have been useful.

I think the only exceptions might either be agent names related to the lexical non-determinism warning during the build, or agent names with non-ASCII chars -- they might actually be illegal labels. It's been a few years since I've looked into this.

I've actually used this in the past for more elaborate migration of jobs between static, manually and heavily customized agents. Basically, have jobs tied to an agent `snowflake`, have an agent `showflake-new`, give it the `snowflake` label, mark `snowflake` as offline, and see what happens. (Which leads to fun label expressions like `snowflake && !snowflake-new` for jobs that don't yet work on the new machine…)